### PR TITLE
Doubleclick Fast Fetch: Measure potential delay of waiting for all ad urls to be constructed

### PR DESCRIPTION
**Pull request by keithwrightbos**
In advance of adding Doubleclick Fast Fetch SRA support (see #9115), measure the amount of time between first ad slot url construction and last.  This will help determine if explicit SRA (waiting for each slot to build its URL) is viable by determining the amount of delay it imposes on the first slot.  Note:

- Guarded by an experiment "'a4a-measure-get-ad-urls"
- Only enabled for doubleclick/adsense Fast Fetch based and CSI sampling indicates data collection
- Results in sending CSI ping

### CODE REVIEW

**Comment by keithwrightbos** [comment]
@dvoytenko mind taking a look?  I still need to create tests.

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 1]
```diff
@@ -680,6 +702,50 @@ export class AmpA4A extends AMP.BaseElement {
...
+           r.element.getAttribute('type') == type;
+       })
+     .then(resources => {
+       if (promiseId != this.promiseId_) {
+         throw cancellation();
+       }
+       const getAdUrlsPromise = [];
+       resources.forEach(r => getAdUrlsPromise.push(r.element.getImpl().then(
+         instance => {
+           //
```
Comment (dvoytenko) [inline extensions/amp-a4a/0.1/amp-a4a.js]: nit: remove

File: src/custom-element.js [hunk 2]
```diff
@@ -1176,6 +1176,14 @@ function createBaseCustomElementClass(win) {
     }
 
     /**
+     * Returns reference to implementation after it has been built.
+     * @return {!Promise<?./base-element.BaseElement>}
```
Comment (dvoytenko) [inline src/custom-element.js]: It'd be `!Promise<!./base-element.BaseElement>` - promise response is non-nullable.

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 1 above)
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Updated to comment:  
// Note that ad url promise could be null if isValidElement returns false.

File: src/custom-element.js (hunk 2 above)
Comment (keithwrightbos) [inline src/custom-element.js]: done

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 3]
```diff
@@ -680,6 +709,54 @@ export class AmpA4A extends AMP.BaseElement {
...
+  measureGetAdUrlsDelay_() {
+    const startTime = Date.now();
+    const type = this.element.getAttribute('type');
+    const promiseId = this.promiseId_;
+    // Only execute once per page, this includes potential pauseCallback flows.
+    if (this.win['a4a-measuring-ad-urls']) {
+      return;
+    }
+    this.win['a4a-measuring-ad-urls'] =
+     (() => resourcesForDoc(this.element).getMeasuredResources(this.win,
```
Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Why are you wrapping this in a closure?

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 4]
```diff
@@ -680,6 +709,54 @@ export class AmpA4A extends AMP.BaseElement {
...
+       r => {
+         return r.element.tagName == 'AMP-AD' &&
+           r.element.getAttribute('type') == type;
+       })
+     .then(resources => {
+       if (promiseId != this.promiseId_) {
+         throw cancellation();
+       }
+       const getAdUrlsPromise = [];
+       resources.forEach(r => getAdUrlsPromise.push(r.element.getImpl().then(
```
Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Use a `#map`.

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 5]
```diff
@@ -680,6 +709,54 @@ export class AmpA4A extends AMP.BaseElement {
...
+         })));
+       return Promise.all(getAdUrlsPromise).then(adUrls => {
+         if (promiseId != this.promiseId_) {
+           throw cancellation();
+         }
+         this.protectedEmitLifecycleEvent_(
+             'sraBuildRequestDelay', {
+               'met.delta.AD_SLOT_ID': Math.round(
+                   this.getNow_() - startTime),
+               'totalSlotCount.AD_SLOT_ID': getAdUrlsPromise.length,
```
Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Nit: use `adUrls`

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 6]
```diff
@@ -680,6 +709,54 @@ export class AmpA4A extends AMP.BaseElement {
...
+      return;
+    }
+    this.win['a4a-measuring-ad-urls'] =
+     (() => resourcesForDoc(this.element).getMeasuredResources(this.win,
+       r => {
+         return r.element.tagName == 'AMP-AD' &&
+           r.element.getAttribute('type') == type;
+       })
+     .then(resources => {
+       if (promiseId != this.promiseId_) {
```
Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Why does a cancellation matter?

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Done

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 6 above)
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Want to ensure we don't send the CSI ping if the page was destroyed.  Its arguably an edge case but see no negative.

Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: 👍

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 4 above)
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Done

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 5 above)
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Done

**Comment by keithwrightbos** [comment]
@jridgewell PTAL, would like to get this into today's cut.  Thanks

**Pull request merged**